File: Selection.vb
Public Class Selection
    Public words As String()
    Public r As Integer ' min length Of word
    Protected complete_group
    Private i As Integer

    Public Function groupSelection(candidates As String(), size As Integer) As List(Of String())
        words = candidates
        r = size
        complete_group = New List(Of String())
        i = 0
        recursiveFillUp(New List(Of String))

        Return complete_group
    End Function

    ' pick elements recursively
    Protected Sub recursiveFillUp(temp As List(Of String))
        Dim picked_elements(words.Length) As List(Of String)
        Dim j As Integer = i
        Do While j < words.Length
            picked_elements(j) = New List(Of String)
            picked_elements(j).AddRange(temp)
            picked_elements(j).Add(words(j))
            ' recoil factor
            If i >= words.Length Then
                i = j
            End If
            ' satisfied yet?
            If picked_elements(j).Count = r Then
                complete_group.Add(picked_elements(j).ToArray())
            ElseIf picked_elements(j).Count < r Then
                recursiveFillUp(picked_elements(j))
            End If
            j += 1
        Loop
        j -= 1
        If picked_elements(j).Equals(Nothing) = False And picked_elements(j).Count = r Then
            i += 1 ' keep recoil factor straightened out
        End If
    End Sub

End Class

File: Selection.cs
using System.Collections.Generic;

namespace Miscellaneous_CS
{
    class Selection
    {
        public string[] words;
        public int r; // min length of word
        protected List<string[]> complete_group;
        private int i;

        public Selection()
        {
        }

        public List<string[]> groupSelection(string[] candidates, int size) {
            words = candidates;
            r = size;
            complete_group = new List<string[]>();
            i = 0;
            recursiveFillUp(new List<string>());

            return complete_group;
        }

        // pick elements recursively
        protected void recursiveFillUp(List<string> temp) {
            List<string>[] picked_elements = new List<string>[words.Length];
            int j = i;
            while (j < words.Length) {
                picked_elements[j] = new List<string>();
                picked_elements[j].AddRange(temp);
                picked_elements[j].Add(words[j]);
                // recoil factor
                if (i >= words.Length) {
                    i = j;
                }
                // satisfied yet?
                if (picked_elements[j].Count == r) {
                    complete_group.Add(picked_elements[j].ToArray());
                }
                else if (picked_elements[j].Count < r) {
                    recursiveFillUp(picked_elements[j]);
                }
                j++;
            }
            if (picked_elements[--j] != null && picked_elements[j].Count == r) {
                i++; // keep recoil factor straightened out
            }
        }
    }
}
